qiskit_experiments/analysis/curve_analysis.py

```diff
@@ -0,0 +1,690 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Analysis class for curve fitting.
+"""
+# pylint: disable=invalid-name
+
+import functools
+from collections import defaultdict
+from typing import Any, NamedTuple, Dict, List, Optional, Tuple, Callable
+
+import numpy as np
+import scipy.optimize as opt
+from qiskit.exceptions import QiskitError
+
+from qiskit_experiments.analysis.data_processing import level2_probability
+from qiskit_experiments.base_analysis import BaseAnalysis
+from qiskit_experiments.data_processing.exceptions import DataProcessorError
+from qiskit_experiments.experiment_data import AnalysisResult, ExperimentData
+
+# Description of data properties for single curve entry
+SeriesDef = NamedTuple(
+    "SeriesDef",
+    [
+        ("name", str),
+        ("param_names", List[str]),
+        ("fit_func_index", int),
+        ("filter_kwargs", Optional[Dict[str, Any]]),
+    ],
+)
+
+# Human readable data set for single curve entry
+CurveEntry = NamedTuple(
+    "CurveEntry",
+    [
+        ("curve_name", str),
+        ("x_values", np.ndarray),
+        ("y_values", np.ndarray),
+        ("y_sigmas", np.ndarray),
+        ("metadata", dict),
+    ],
+)
+
+
+def scipy_curve_fit_wrapper(
+    func: Callable,
+    xdata: np.ndarray,
+    ydata: np.ndarray,
+    p0: np.ndarray,
+    sigma: Optional[np.ndarray],
+    bounds: Optional[Tuple[np.ndarray, np.ndarray]],
+    **kwargs,
+) -> AnalysisResult:
```

Comment:
The `AnalysisResult` is temporary class that will be shortly replaced by `AnalysisResultV1`. This is just a subclass of dictionary, so this is kind of doing what you mentioned :) I'll update the `_run_analysis` to convert it into `AnalysisResultV1` once 5499 is merged.

Reply:
In Jessie's PR this is not a subclass of a dictionary, if I remember correctly

Reply:
Okey, I'll check how we can populate the data.

Reply:
It seems we can pass a free form dictionary to the `result_data` arg of `AnalysisResultV1` constructor. I think `_run_analysis` should create a dictionary and experiment base class will create `AnalysisResultV1` with the dictionary. Because this class requires some other field such as `experiment_id`, `device_components`, etc... the analysis class cannot provide these information.

Perhaps we can rename current `AnalysisResult` to `ResultData` to conform to the terra API. Basically this is just a wrapper class of dictionary so we can directly pass the object to `result_data`. This is convenient for type check.

Reply:
> It seems we can pass a free form dictionary to the result_data arg of AnalysisResultV1 constructor.

I agree.

> Because this class requires some other field such as experiment_id, device_components, etc... the analysis class cannot provide these information.

I think essentially this information is available in `_run_analysis` from the experiment data and the circuits metadata (at least for T1 I managed to do it, see #29). However I don't like these inexplicit ways of passing information from the experiments class to the analysis class.

> Basically this is just a wrapper class of dictionary so we can directly pass the object to `result_data`.

What's the advantage of having an `AnalysisResult` class inheriting from `dict`, over just using `dict`?

Reply:
I see. Thanks. The data field necessary to populate `AnalysisResultV1 ` is always the same? Of is this subclass dependent?

> What's the advantage of having an AnalysisResult class inheriting from dict, over just using dict?

Just for type hint. If the hint is always `Dict[str, Any]` we need to read the code of other method to check what is passed.

Reply:
What do you mean by "The data field necessary to populate AnalysisResultV1 is always the same? Of is this subclass dependent?" ?

Reply:
I meant common dict keys. Because `result_data` will be shown in the IQX, I guess we should define some necessary fields in the `result_data`. Or is this completely free form dict and IQX is smart enough to understand what is descried in the dictionary?

Reply:
I don't know; @jyu00 @chriseclectic